opnode/rollup/driver/state.go

```diff
@@ -196,24 +217,42 @@ func (s *state) createNewL2Block(ctx context.Context) (eth.L1BlockRef, error) {
 		s.log.Error("Error finding next L1 Origin", "err", err)
 		return eth.L1BlockRef{}, err
 	}
-	if nextOrigin.Time <= s.Config.BlockTime+s.l2Head.Time {
-		s.log.Trace("Skipping block production because the next block time is behind the next L1 Origin", "l2Head", s.l2Head, "l1Origin", nextOrigin)
+	nextL2Time := s.l2Head.Time + s.Config.BlockTime
+	// If we are behind the L1 origin that we should be using then we broke the invariant
+	if nextL2Time < nextOrigin.Time {
+		s.log.Error("Cannot build L2 block for time before L1 origin",
+			"l2Head", s.l2Head, "nextL2Time", nextL2Time, "l1Origin", nextOrigin, "l1OriginTime", nextOrigin.Time)
+		return eth.L1BlockRef{}, fmt.Errorf("cannot build L2 block on top %s for time %d before L1 origin %s at time %d",
+			s.l2Head, nextL2Time, nextOrigin, nextOrigin.Time)
+	}
+
+	// If the L1 origin changed this block, then we are in the first block of the epoch
+	isFirstEpochBlock := s.l2Head.L1Origin.Number != nextOrigin.Number
+
+	// We create at least 1 block per epoch. After that we have to enforce the max timestamp.
+	//
+	// TODO: we can give ourselves less slack here, and prefer to adopt the next L1 block more eagerly,
+	// to ensure we have more time to submit the next block we produce, at the cost of sequencing less eagerly.
+	maxL2Time := nextOrigin.Time + s.Config.MaxSequencerDrift
```

Comment:
This still have the same problem that it will not fill out blocks up to the next L1 block if it instead should be. This is not consensus critical, but still slightly odd.

Reply:
Ah, forgot we were checking it there again. We should allow block creation, the verifier will accept the blocks, will update this.

Reply:
[8393ced](https://github.com/ethereum-optimism/optimistic-specs/pull/315/commits/8393ced722e5f15f65f964e6fecad9be2b45d43c) should fix this

Reply:
lgtm